src/hashing.rs

```diff
@@ -0,0 +1,495 @@
+/// 1. The version number is fixed for a given schema of the domain tag.
+/// 2. The domain represents a broad class of use cases for the hashing, e.g. MACs, or key derivation.
+/// 3. The label is used to differentiate different applications of a use case. e.g. there might be two places key
+/// derivation is used in your application: for wallet derived keys, and communication derived keys. These might have
+/// the label "wallet-key" and "comms-key" respectively.
+///
+/// [`DomainSeparatedHasher`] is useful for more generic use-cases that aren't covered by the two primary use cases
+/// covered in this API (MAcs and key derivation).
+///
+/// ## Examples
+///
+/// Using a hash as an object ID, based on the fields of the object.
+///
+/// ```
+/// use sha2::Sha256;
+/// use tari_crypto::hashing::{DomainSeparatedHash, DomainSeparatedHasher, GenericHashDomain};
+/// use tari_utilities::hex::{to_hex, Hex};
+/// struct Card {
+///     name: &'static str,
+///     strength: u8,
+/// }
+///
+/// fn card_id(card: &Card) -> DomainSeparatedHash {
+///     DomainSeparatedHasher::<Sha256, GenericHashDomain>::new("card_id")
+///         .chain(card.name.as_bytes())
+///         .chain(&[card.strength])
+///         .finalize()
+/// }
+///
+/// let card = Card {
+///     name: "Rincewind",
+///     strength: 8,
+/// };
+///
+/// let id = card_id(&card);
+/// assert_eq!(id.domain_separation_tag(), "com.tari.generic.v1.card_id");
+/// assert_eq!(
+///     to_hex(id.as_ref()),
+///     "44fb39bfdd20c93ddf542e4b2d1f4b06448aa1fa2b9c4d138d8e7bbb19aa7c65"
+/// );
+/// ```
+///
+/// Calculating a signature challenge
+///
+/// ```
+/// use tari_crypto::{
+///     common::Blake256,
+///     hashing::{DomainSeparatedHash, DomainSeparatedHasher, GenericHashDomain},
+/// };
+/// use tari_utilities::hex::{to_hex, Hex};
+/// struct Card {
+///     name: &'static str,
+///     strength: u8,
+/// }
+///
+/// fn calculate_challenge(msg: &str) -> DomainSeparatedHash {
+///     DomainSeparatedHasher::<Blake256, GenericHashDomain>::new("schnorr_challenge")
+///         .chain(msg.as_bytes())
+///         .finalize()
+/// }
+///
+/// let challenge = calculate_challenge("All is well.");
+/// assert_eq!(
+///     challenge.domain_separation_tag(),
+///     "com.tari.generic.v1.schnorr_challenge"
+/// );
+/// assert_eq!(
+///     to_hex(challenge.as_ref()),
+///     "6cd8efe7d7f1673ed8cfc0ac67d6979eb50afdbf276adf5221caabfbfd01da8c"
+/// );
+/// ```
+pub struct DomainSeparatedHasher<D, M> {
+    inner: D,
+    label: String,
+    dst: PhantomData<M>,
+}
+
+impl<D: Digest, M: DomainSeparation> DomainSeparatedHasher<D, M> {
+    pub fn new<S>(label: S) -> Self
+    where S: AsRef<str> {
+        let inner = D::new();
+        let mut result = Self {
+            inner,
+            label: label.as_ref().to_string(),
+            dst: PhantomData::<M>::default(),
+        };
+        result.update(M::domain_separation_tag(&label).as_bytes());
```

Comment:
@CjS77  @AaronFeickert made this comment in one of the discussions:

> Provided the method used for including the label and message in D is secure (simple concatenation, for example, is not sufficient), D_1 and D_2 behave as independent high-quality cryptographic hash functions, and generally retain the useful properties of D

I may be misunderstanding, but aren't we "simply concatenating (prepending)" the label to the message? Personalization in the Blake2b crate seems to be accomplished by passing a `persona` to `with_params`.

for e.g. 
```rust
VarBlake2b::with_params(
  //  key
  &[],
  // salt
  &[]
  // persona - must be 32 bytes
  b"com.tari.custom-32-byte-string_"
)
```

Or is this not required?

Reply:
Safe label prepending is intended to be equivalent in functionality to the `Blake2b` `persona` parameter, but not every hash function's common APIs support this natively.

What I meant by "simple concatenation" being insufficient is that without proper handling of variable-length input, you can obtain collisions. Consider two examples that use concatenation
```
label_1 = "cats"
m_1 = "dogs"

label_2 = "cat"
m_2 = "sdogs"

D(label_1 || m_1) = D(label_2 || m_2) = D("catsdogs") // collision!
```

Reply:
In the API that @CjS77 wrote, each label's length is prepended before the label itself, so there's no doubt where the label ends and subsequent inputs begin.

Reply:
Ok thanks for clearing that up - makes sense